### ISSUE TITLE
GMNN and baselines run with different features, that is not fair and proper.

### ISSUE THREAD

**Issue by yaokl-nju** [DeepGraphLearning/GMNN#2]
Hello guys,
  I found that GMNN runs with the binarized features, but other baselines run with the real-valued features. I think that the experimental results are not fair and proper.

  Also, I run GCN with the binarized features, and it also can achieve good results. Under this settings, the margins between GCN and GMNN is very small.

  Besides, I run GMNN with real-valued features, and the performances drop a lot.

**Comment by yaokl-nju** [comment]
Besides above, it is weird that different optimizers are used for different datasets.

**Comment by mnqu** [comment]
Hi klyao,

Thanks for pointing it out!

On Cora and Citeseer, the raw node features were already binarized. For Pubmed, when tuning our models, we occasionally found that our model can achieve better results by using binary-valued features, so we uniformly used binary-valued features in our code base for all the datasets, as you pointed out.

For your question:
1. On Cora and Citeseer in table 2, since the raw node features were already binarized, all the compared algorithms used binary-valued features in these cases. As a result, the settings of all algorithms were almost the same, and GMNN significantly outperformed all the baseline algorithms. On the Pubmed dataset in table 2, we directly reported the results of GCN and GAT from the original papers, which was a common practice. Therefore, we didn't run them with our own code base, which may cause some difference on how the features were used, since the raw node features of Pubmed were real-valued.
2. To get a more proper comparison on Pubmed in table 2 as you suggest, we rerun GCN with binary-valued features. The mean accuracy over 100 runs is 79.9%, and the margin from GMNN (81.8%) is still very large instead of very small. Moreover, we also rerun GMNN with real-valued features on Pubmed, and the mean accuracy is 80.5% over 100 runs, which is still one of the SOTA results on Pubmed. For this experiment, we use the simplest variant of GMNN (i.e., GMNN W/o Attr. in p). The RMSprop optimizer is used, with the initial learning rate as 0.05 and weight decay as 0.0005. The result can be easily reproduced by slightly changing the running script (run_pubmed.py), and also changing the line 69 of train.py as _feature.to_one_hot(binary=False)_. In addition, by using other GMNN variants (e.g., GMNN With Attr. in p), we should be able to further improve the results. For example, it is easy to improve the accuracy up to 80.8% over 100 runs by slightly tuning the model.
3. In the appendix of the paper, we ran all the compared algorithms (i.e., GCN, GAT, GMNN) by using our own code base, where binary-valued features were used for all algorithms on all datasets. In this case, the experimental settings of different algorithms were almost the same, and GMNN significantly outperformed GCN and GAT under random data splits and few-shot learning settings on Pubmed.
4. On Pubmed, the Adam optimizer was used since it was a little bit more stable in this case. We can also use RMSprop, which will achieve almost the same results.

Due to the limited space, we didn't describe some experimental details very clearly in the previous paper, and we will make it clearer in the updated version.

Hope that our reply can help. Finally, thank you again for pointing out the issue and helping improve our paper!

Meng

**Comment by yaokl-nju** [comment]
Hi Meng,
 
  Thanks for your patient reply!

  Although it looks fair now and it looks fair to copy the results of GCN from the original paper, we found that GCN can achieve better results. Actually, I can run GCN with dim=64 and get results of cora=0.832, citeseer=0.727, pubmed=0.803 with binary-valued feature (100 runs).

  I know it is common for most papers that they directly copy results of GCN from the original paper, but I don't think that it is reasonable and good for this research area.

  Anyway, thanks for your patient reply and your work is great!

**Comment by mnqu** [comment]
Hi klyao,

Thank you for the valuable feedback!

It is quite surprising that GCN can achieve such good results. In our paper, the hidden dimension of both GCN and GMNN was fixed to 16, as suggested in the original GCN paper. With the hyperparameters you provided, the results of GCN are indeed improved. Besides, the results of GMNN are also improved with this hyperparameter. For example, on Cora, the simplest GMNN variant now achieves an accuracy of 83.8% over 100 runs.

I agree with you that every researcher in this field should run GCN more carefully. In the future, we will try to further tune the hyperparameters of GCN and GMNN. Hopefully, the results of both models can be further improved.

Overall, thank you for the valuable discussion and comments! : )

Meng
